src/gt4py/analysis/passes.py

```diff
@@ -865,3 +865,133 @@ def _make_axis_interval(self, interval: IntervalInfo):
         result = gt_ir.AxisInterval(start=axis_bounds[0], end=axis_bounds[1])
 
         return result
+
+
+class DemoteLocalTemporariesToVariablesPass(TransformPass):
+    class CollectDemotableSymbols(gt_ir.IRNodeVisitor):
+        def __call__(self, node):
+            assert isinstance(node, gt_ir.StencilImplementation)
+            self.demotables = set(node.temporary_fields)
+            self.seen_symbols = set()
+            self.stage_symbols = None
+            self.visit(node)
+            return self.demotables
+
+        def visit_Stage(self, node: gt_ir.Stage):
+            self.stage_symbols = set()
+
+            self.generic_visit(node)
+
+            for s in self.stage_symbols:
+                if s in self.seen_symbols:
+                    self.demotables.discard(s)
+                self.seen_symbols.add(s)
+            self.stage_symbols = None
+
+        def visit_FieldRef(self, node: gt_ir.FieldRef):
+            if any(v != 0 for v in node.offset.values()):
+                self.demotables.discard(node.name)
+            self.stage_symbols.add(node.name)
+
+    class DemoteSymbols(gt_ir.IRNodeMapper):
```

Comment:
Just a comment: this is the first time we use the `NodeMapper` in a transformation, right?

Reply:
Indeed, that's why I had to fix it :smile: